### PR TITLE
Knn Col-Major

### PR DESCRIPTION
**Pull request by KulikovNikita**

### CODE REVIEW

File: cpp/oneapi/dal/backend/primitives/search/test/l2_search_dpc.cpp [hunk 1]
```diff
@@ -37,11 +37,52 @@ namespace te = dal::test::engine;
...
+struct order_map<f_order> {
+    constexpr static auto value = ndorder::f;
+};
+
+template <typename order>
+constexpr auto order_v = order_map<order>::value;
+
+template <typename T, ndorder order>
+auto table_to_ndarray(sycl::queue& q, const table& t) {
+    const auto rc = t.get_row_count();
```
Comment (amgrigoriev) [inline cpp/oneapi/dal/backend/primitives/search/test/l2_search_dpc.cpp:61]: Can you make it public to enable re-usage?

File: cpp/oneapi/dal/backend/primitives/utils.hpp [hunk 2]
```diff
@@ -38,12 +39,109 @@ inline ndarray<Type, 1> table2ndarray_1d(const table& table) {
 
 #ifdef ONEDAL_DATA_PARALLEL
 template <typename Type>
-inline ndarray<Type, 2> table2ndarray(sycl::queue& q,
-                                      const table& table,
-                                      sycl::usm::alloc alloc = sycl::usm::alloc::shared) {
+inline ndarray<Type, 2, ndorder::c> table2ndarray_rm(sycl::queue& q,
+                                                     const table& table,
```
Comment (amgrigoriev) [inline cpp/oneapi/dal/backend/primitives/utils.hpp:43]: does it work for non-homogen table?

File: cpp/oneapi/dal/backend/primitives/distance/lp_distance_dpc.cpp [hunk 3]
```diff
@@ -18,23 +18,105 @@
 
 namespace oneapi::dal::backend::primitives {
 
+template <typename Float, ndorder ord, typename Idx>
+struct dkeeper {};
+
+template <typename Float, typename Idx>
+struct dkeeper<Float, ndorder::c, Idx> {
+    struct row_iterator {
+        const Float& operator*() const {
```
[the remark this answers is not in the document]
Comment (KulikovNikita) [reply cpp/oneapi/dal/backend/primitives/distance/lp_distance_dpc.cpp:27]: Will be done

File: cpp/oneapi/dal/backend/primitives/distance/lp_distance_dpc.cpp [hunk 4]
```diff
@@ -18,23 +18,105 @@
...
+            return *(row + col);
+        }
+
+        row_iterator& operator++() {
+            ++col;
+            return *this;
+        }
+
+        bool operator!=(const row_iterator& rhs) const {
+            const auto& lhs = *this;
```
[the remark this answers is not in the document]
Comment (KulikovNikita) [reply cpp/oneapi/dal/backend/primitives/distance/lp_distance_dpc.cpp:37]: There is no way to check it

File: cpp/oneapi/dal/backend/primitives/distance/lp_distance_dpc.cpp [hunk 5]
```diff
@@ -18,23 +18,105 @@
...
+
+template <typename Float, typename Idx>
+struct dkeeper<Float, ndorder::c, Idx> {
+    struct row_iterator {
+        const Float& operator*() const {
+            return *(row + col);
+        }
+
+        row_iterator& operator++() {
+            ++col;
```
[the remark this answers is not in the document]
Comment (KulikovNikita) [reply cpp/oneapi/dal/backend/primitives/distance/lp_distance_dpc.cpp:32]: Discussed

File: cpp/oneapi/dal/backend/primitives/search/test/l2_search_dpc.cpp [hunk 6]
```diff
@@ -37,11 +37,52 @@ namespace te = dal::test::engine;
 namespace de = dal::detail;
 namespace la = te::linalg;
 
-template <typename Float>
-class search_test : public te::float_algo_fixture<Float> {
+class c_order {};
```
[the remark this answers is not in the document]
Comment (KulikovNikita) [reply cpp/oneapi/dal/backend/primitives/search/test/l2_search_dpc.cpp:40]: discussed

File: cpp/oneapi/dal/backend/primitives/distance/lp_distance_dpc.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (KulikovNikita) [reply cpp/oneapi/dal/backend/primitives/distance/lp_distance_dpc.cpp:33]: Fixed

**Comment by KulikovNikita** [comment]
Old CI: http://intel-ci.intel.com/ec751038-27ed-f1c5-9def-8cdcd4b723a1

**Comment by KulikovNikita** [comment]
Manually started CI:  http://intel-ci.intel.com/ec753d0c-4cfe-f1cc-aa3e-8cdcd4b723a1

**Review by KalyanovD — APPROVED** [review]

**Comment by KalyanovD** [comment]
image

**Comment by KulikovNikita** [comment]
Yet another one manually started CI: Manually started CI: http://intel-ci.intel.com/ec77669e-3dea-f107-a6e9-8cdcd4b723a1

**Comment by KulikovNikita** [comment]
Manually started CI: http://intel-ci.intel.com/ec784bc1-b6b6-f195-a2fb-8cdcd4b723a1

**Pull request merged**